drv/cosmo-seq-server/src/vcore.rs

```diff
@@ -0,0 +1,455 @@
+    EreportSentOff(Rail, usize),
+    EreportTooBig(Rail),
+}
+
+#[derive(Copy, Clone, PartialEq)]
+pub struct Rails {
+    pub vddcr_cpu0: bool,
+    pub vddcr_cpu1: bool,
+}
+
+ringbuf!(Trace, 60, Trace::None);
+
+///
+/// We are going to set our input undervoltage warn limit to be 11.75 volts.
+/// Note that we will not fault if VIN goes below this (that is, we will not
+/// lose POWER_GOOD), but the part will indicate an input fault and pull
+/// on its PMBus alert pin.
+///
+const VCORE_UV_WARN_LIMIT: units::Volts = units::Volts(11.75);
+
+///
+/// We want to collect enough samples (at ~900µs per sample per regulator, or
+/// ~1.8ms for both) to adequately cover any anticipated dip.  We have seen
+/// these have an ~11ms total width in the wild, so we give ourselves plenty
+/// of margin here and get ~45ms of data.
+///
+/// (Read: I just took the number Bryan picked in the Gimlet sequencer,
+/// divided it by 2, and copied his comment lol)
+///
+const VCORE_NSAMPLES: usize = 25;
+
+impl VCore {
+    pub fn new(i2c: TaskId, packrat: task_packrat_api::Packrat) -> Self {
+        let (device, rail) = i2c_config::pmbus::vddcr_cpu0_a0(i2c);
+        let vddcr_cpu0 = Raa229620A::new(&device, rail);
+
+        let (device, rail) = i2c_config::pmbus::vddcr_cpu1_a0(i2c);
+        let vddcr_cpu1 = Raa229620A::new(&device, rail);
+        Self {
+            vddcr_cpu0,
+            vddcr_cpu1,
+            packrat,
+        }
+    }
+
+    pub fn initialize_uv_warning(&self) -> Result<(), ResponseCode> {
+        ringbuf_entry!(Trace::Initializing);
+
+        // Set our warn limit
+        retry_i2c_txn(Rail::VddcrCpu0, PmbusCmd::LoadLimit, || {
+            self.vddcr_cpu0.set_vin_uv_warn_limit(VCORE_UV_WARN_LIMIT)
+        })?;
+        retry_i2c_txn(Rail::VddcrCpu1, PmbusCmd::LoadLimit, || {
+            self.vddcr_cpu1.set_vin_uv_warn_limit(VCORE_UV_WARN_LIMIT)
+        })?;
+        ringbuf_entry!(Trace::LimitsLoaded);
+
+        // Clear our faults
+        self.clear_faults(Rails {
+            vddcr_cpu0: true,
+            vddcr_cpu1: true,
+        })?;
+
+        // The higher-level sequencer code will unmask the FPGA interrupts for
+        // our guys.
+
+        ringbuf_entry!(Trace::Initialized);
+
+        Ok(())
+    }
+
+    pub fn clear_faults(&self, which_rails: Rails) -> Result<(), ResponseCode> {
+        if which_rails.vddcr_cpu0 {
+            retry_i2c_txn(Rail::VddcrCpu0, PmbusCmd::ClearFaults, || {
+                self.vddcr_cpu0.clear_faults()
+            })?;
+        }
+
+        if which_rails.vddcr_cpu1 {
+            retry_i2c_txn(Rail::VddcrCpu1, PmbusCmd::ClearFaults, || {
+                self.vddcr_cpu1.clear_faults()
+            })?;
+        }
+
+        ringbuf_entry!(Trace::FaultsCleared(which_rails));
+
+        Ok(())
+    }
+
+    pub fn handle_pmbus_alert(&self, mut rails: Rails, now: u64) {
+        ringbuf_entry!(Trace::PmbusAlert {
+            timestamp: now,
+            rails,
+        });
+
```

Comment:
Mutation of `rails` is a little weird here.  When we call the function, it represents "rails which have triggered an FPGA alert", but then we set the flags for rails that have a fault bit set.  We don't clear the flags if the rails **don't** have the fault bit set, and we don't use the incoming flags to decide which rails to check.

If a rail has triggered an FPGA alert but _doesn't_ actually have a fault bit set, we will erroneously try to clear its fault when calling `clear_faults`.

I'm not sure how this is supposed to go; maybe we shouldn't pass it in at all?

Reply:
tried to explain it a little better in https://github.com/oxidecomputer/hubris/pull/2184/commits/238ff27f36aaed5755a78847de562d5678b4efbe